### DISCUSSION TITLE
Type problem: "Types of property 'register' are incompatible. Type 'UseFormRegister<FormData>' is not assignable to type 'UseFormRegister<Record<string, any>>'"

### DISCUSSION THREAD

**Discussion by shackra** [react-hook-form/react-hook-form#5868] [Q&A]
Hi!

I'm experiencing a weird type problem, I think is related to the usage of typing for [react-select](https://react-select.com/home)'s Select control:

```
Type 'Control<FormData>' is not assignable to type 'Control<Record<string, any>>'.
  Types of property 'register' are incompatible.
    Type 'UseFormRegister<FormData>' is not assignable to type 'UseFormRegister<Record<string, any>>'.
      Type 'Record<string, any>' is not assignable to type 'FormData'.ts(2322)
controller.d.ts(22, 5): The expected type comes from property 'control' which is declared here on type 'IntrinsicAttributes & { render: ({ field, fieldState, formState, }: { field: ControllerRenderProps<Record<string, any>, `${string}` | `${string}.${string}` | `${string}.${number}`>; fieldState: ControllerFieldState; formState: FormState<...>; }) => ReactElement<...>; } & UseControllerProps<...>'
```

Here is the codesandbox link https://codesandbox.io/s/rhf-issue-with-the-control-prop-type-mx9re?file=/src/App.tsx

**Comment by shackra** [comment]
hi!

any help? is this a bug worth reporting?

**Reply by bluebill1049** [reply]
your CSB is broken, I think the issue is because react-select is a group input.  you may want to use this type: https://react-hook-form.com/ts#NestedValue

**Reply by shackra** [reply]
> your CSB is broken, I think the issue is because react-select is a group input. you may want to use this type: https://react-hook-form.com/ts#NestedValue

~not sure how to fix my CSB.~ somehow, `field.ref` was at fault if passed to react-select Select component in its `ref` prop.

I try using `NestedValue` but react-select has `null` in the union type:

```
Type '{ value: number; label: string; } | null' does not satisfy the constraint 'unknown[] | Record<string, unknown> | Map<unknown, unknown>'.
  Type 'null' is not assignable to type 'unknown[] | Record<string, unknown> | Map<unknown, unknown>'.
```

**Reply by shackra** [reply]
[even if I don't use react-select's `ValueType` generic and use `NestedValue`](https://codesandbox.io/s/rhf-issue-with-the-control-prop-type-forked-1jn3p?file=/src/App.tsx), the error is still there:

```
(JSX attribute) control?: Control<Record<string, any>> | undefined
Type 'Control<FormData>' is not assignable to type 'Control<Record<string, any>>'.
  Types of property 'register' are incompatible.
    Type 'UseFormRegister<FormData>' is not assignable to type 'UseFormRegister<Record<string, any>>'.
      Type 'Record<string, any>' is not assignable to type 'FormData'.ts(2322)
controller.d.ts(22, 5): The expected type comes from property 'control' which is declared here on type 'IntrinsicAttributes & { render: ({ field, fieldState, formState, }: { field: ControllerRenderProps<Record<string, any>, `${string}` | `${string}.${string}` | `${string}.${number}`>; fieldState: ControllerFieldState; formState: FormState<...>; }) => ReactElement<...>; } & UseControllerProps<...>'
```

**Reply by shackra** [reply]
Well, I did another change in hopes to figure out what's wrong, and discovered that the values pass to the submission function come as:

![image](https://user-images.githubusercontent.com/1055216/125553813-6e64a895-b3a1-4d4b-bb8d-d90e36613bcf.png)

An `undefined` key containing the values from react-select's Select!

![image](https://user-images.githubusercontent.com/1055216/125554143-39376ccd-1b2a-437c-9888-f86c8e99422b.png)

why would `name` and `value` come as `undefined` from the `field` of Controller?

**Comment by shackra** [comment]
nvm, turns out omitting `name` prop on the Controller component causes this issue
